diff --git a/utils/rtm-naming/Makefile.am b/utils/rtm-naming/Makefile.am
index a9e7bb5e..151a2ed6 100644
--- a/utils/rtm-naming/Makefile.am
+++ b/utils/rtm-naming/Makefile.am
@@ -13,10 +13,10 @@ ORB = @ORB@
 bin_SCRIPTS = rtm-naming
 
 rtm-naming: rtm-naming.in
-	echo '#!/bin/sh' > $@
+	echo '#!/bin/bash' > $@
 	echo '' >> $@
-	echo 'cosnames="'${COSNAMES}'"' >> $@
-	echo 'orb="'${ORB}'"' >> $@
+	echo 'COSNAMES="'${COSNAMES}'"' >> $@
+	echo 'ORB="'${ORB}'"' >> $@
 	cat rtm-naming.in >> $@
 	chmod 755 $@
 
diff --git a/utils/rtm-naming/rtm-naming.in b/utils/rtm-naming/rtm-naming.in
index d9c586a8..794c8350 100644
--- a/utils/rtm-naming/rtm-naming.in
+++ b/utils/rtm-naming/rtm-naming.in
@@ -13,10 +13,11 @@
 #  $Id$
 #
 
-pname=`basename $cosnames`
-hostname=`hostname`
-currdir=`pwd`
-DEBUG=""
+PNAME=`basename $COSNAMES`
+HOSTNAME=`hostname`
+OSNAME=`uname -s`
+LOGDIR="/tmp"
+DEBUG="YES"
 
 usage()
 {
@@ -41,6 +42,7 @@ EOF
 
 get_opt()
 {
+    debug "get_opt()"
     # Global variables
     FORCEKILL=""
     PASSWORD=""
@@ -65,6 +67,11 @@ get_opt()
                 ;;
         esac
     done
+    debug "Option related variables:"
+    debug "  PORT " $PORT
+    debug "  FORCEKILL " $FORCEKILL
+    debug "  STOP" $STOPNAMESERVICE
+    debug "  PASSWORD" $PASSWORD
 }
 #
 # debug <text>
@@ -76,7 +83,7 @@ get_opt()
 debug()
 {
     if test "x$DEBUG" != "x"; then
-        echo $*
+        echo $* >> /tmp/rtm-naming-${USER}_debug.log
     fi
 }
 
@@ -89,9 +96,15 @@ debug()
 # @return 0 process that is using the port was found
 #         1 process not found
 #
+# Linux: netstat -tlnp
+#        t: TCP
+#        l: listening
+#        n: numeric IP address
+#        p: print program and PID
 #
 get_pid_of_nsport()
 {
+    debug "get_pid_of_nsport()"
     _port=$1
     debug "netstat -tlnp 2> /dev/null | grep $_port | awk '{print $7;}'"
     netstat_tanp=`netstat -tlnp`
@@ -113,7 +126,43 @@ get_pid_of_nsport()
     debug "$_port port is used by $pname_of_nsport (pid = $_pid)."
     return 0
 }
+#
+# get_pid_of_nsport <port_number> for macOS
+#
+# @param  pid_of_nsport: Process ID of a process which is using the port.
+# @param  pname_of_nsport: Process name which is using the port is set
+# @param  pid_of_nsport: Process ID which is using the port is set
+# @return 0 process that is using the port was found
+#         1 process not found
+#
+# macOS: lsof
+#        -n: numeric IP address (not hostname)
+#        -P: numeric Port number (not service name)
+#        -i: specifying address/port, :<port> specifies port
+#
+get_pid_of_nsport_macos()
+{
+    debug "get_pid_of_nsport_macos()"
+    _port=$1
+    debug "sudo -S lsof -n -P -i :$_port | grep LISTEN"
+    if test "x$PASSWORD" = "x"; then
+        _netstat=`sudo -S lsof -n -P -i :$_port | grep LISTEN`
+    else
+        _netstat=`echo $PASSWORD | sudo -S lsof -n -P -i :$_port | grep LISTEN`
+    fi
+    debug "A possible process that is using $_port port:" $_netstat
+    if test "x$_netstat" = "x"; then
+        echo "No process using port number ${_port} on the system."
+        debug "No process using port number ${_port} on the system."
+        return 1
+    fi
 
+    _pid=`echo $_netstat | awk '{print $2;}'`
+    pname_of_nsport=`echo $_netstat | awk '{print $1;}'`
+    pid_of_nsport=$_pid
+    debug "$_port port is used by $pname_of_nsport (pid = $_pid)."
+    return 0
+}
 
 #
 # is_launch_from_init <prog_name> (<pid>)
@@ -148,8 +197,8 @@ is_started_from_init()
 	if test $_pid -eq $_var_run_pid; then
 	    debug "The process $_pname ($_pid) might be started " \
 		"from init script."
-	    debug "searching init script: $cosnames"
-	    init_script=`grep $cosnames /etc/init.d/* | awk 'BEGIN{FS=":";}{if(FNR==1){print $1;}}'`
+	    debug "searching init script: $COSNAMES"
+	    init_script=`grep $COSNAMES /etc/init.d/* | awk 'BEGIN{FS=":";}{if(FNR==1){print $1;}}'`
 
 	    if test "x$init_script" = "x"; then
 
@@ -176,8 +225,9 @@ is_started_from_init()
 #
 check_cosname()
 {
-    if test ! -f $cosnames ; then
-	echo "Name service program ($cosnames) not found."
+    debug "check_cosname()"
+    if test ! -f $COSNAMES ; then
+	echo "Name service program ($COSNAMES) not found."
 	echo "Please install or chech rtm-naming script."
 	exit 1
     fi
@@ -185,16 +235,31 @@ check_cosname()
 
 delete_omninames_files()
 {
+    debug "delete_omninames_files()"
+    _logfiles=(
+        $LOGDIR/omninames-$HOSTNAME.log
+        $LOGDIR/omninames-$HOSTNAME.bak
+        $LOGDIR/omninames-$HOSTNAME.dat
+        )
     if test "x$PASSWORD" = "x" ; then
-        rm -f ./omninames-$hostname.log
-        rm -f ./omninames-$hostname.bak
-        rm -f ./omninames-$hostname.dat
+        for file in ${_logfiles[@]} ; do
+            if test -w $file ; then
+                rm -f $file
+            else
+                debug "log files are not writable. delete by sudo."
+                sudo -S rm -f $file
+            fi
+        done
     else
-        echo $PASSWORD | sudo -S rm -f ./omninames-$hostname.log
-        echo $PASSWORD | sudo -S rm -f ./omninames-$hostname.bak
-        echo $PASSWORD | sudo -S rm -f ./omninames-$hostname.dat
+        for file in ${_logfiles[@]} ; do
+            echo $PASSWORD | sudo -S rm -f $file
+        done
+    fi
+    if test $? != 0 ; then
+        echo "Delete log files failed!!"
+        debug "Delete log files failed!!"
     fi
-    debug "omninames-$hostname files deleted."
+    debug "omninames-$HOSTNAME files deleted."
 }
 
 #
@@ -204,12 +269,17 @@ delete_omninames_files()
 #
 specified_port_used_check()
 {
-    get_pid_of_nsport $PORT
+    debug "specified_port_used_check()"
+    if test "x$OSNAME" = "xDarwin" ; then
+        get_pid_of_nsport_macos $PORT
+    else
+        get_pid_of_nsport $PORT
+    fi
     if test $? -eq 0; then
         debug "The Process information using the port could be obtained."
         # If "port" is used by other program -> abort
 
-        pids=`pgrep $pname`
+        pids=`pgrep $PNAME`
         matchflag=0
         for p in $pids; do
             if test "x$pid_of_nsport" = "x$p"; then
@@ -217,8 +287,8 @@ specified_port_used_check()
             fi
         done
         if test $matchflag -eq 0; then
-            echo "$pname_of_nsport (not $pname) is using the port."
-            echo "Starting $pname aborted. Please use the other port."
+            echo "$pname_of_nsport (not $PNAME) is using the port."
+            echo "Starting $PNAME aborted. Please use the other port."
             exit 1
         fi
     else
@@ -235,6 +305,7 @@ specified_port_used_check()
 #
 stop_omninames_by_init_script()
 {
+    debug "stop_omninames_by_init_script()"
     echo "omniNames might be started $init_script."
 
     if test "x$FORCEKILL" = "x" && test "x$STOPNAMESERVICE" = "x" ; then
@@ -248,13 +319,13 @@ stop_omninames_by_init_script()
     echo "Stopping omniNames by $init_script."
     if test "x$PASSWORD" = "x" ; then
        sudo $init_script stop
-       sudo rm -f /var/run/$pname.pid
+       sudo rm -f /var/run/$PNAME.pid
     else
         echo $PASSWORD | sudo -S $init_script stop
-        echo $PASSWORD | sudo -S rm -f /var/run/$pname.pid
+        echo $PASSWORD | sudo -S rm -f /var/run/$PNAME.pid
     fi
     debug "$init_script stop"
-    debug "/var/run/$pname.pid  are deleted"
+    debug "/var/run/$PNAME.pid  are deleted"
     delete_omninames_files
     return 0
 
@@ -278,7 +349,8 @@ stop_omninames_by_init_script()
 #
 stop_existing_ns()
 {
-    debug "---stop_existing_ns"
+    debug "stop_existing_ns()"
+    debug "---stop_existing_ns---"
 
     # Find a process which is using specified port.
     specified_port_used_check
@@ -288,14 +360,14 @@ stop_existing_ns()
 
     # omniNames is not started by init script
     if test $? -ne 0; then
-        debug "$pname might not be started from init script."
+        debug "$PNAME might not be started from init script."
         if test "x$pid_of_nsport" != "x" ; then
-            echo "$pname (pid: $pid_of_nsport) is running"
+            echo "$PNAME (pid: $pid_of_nsport) is running"
 
             # Restart naming service
             if test "x$FORCEKILL" = "x" ; then
                 if test "x$STOPNAMESERVICE" = "x" ; then
-		            read -p "Kill anyway and start $pname again? (y/N)" killns
+		            read -p "Kill anyway and start $PNAME again? (y/N)" killns
 	            else
                     read -p "Kill anyway? (y/N)" killns
 		        fi
@@ -305,16 +377,18 @@ stop_existing_ns()
                 fi
 	        fi
             if test "x$PASSWORD" = "x" ; then
+                debug "Killing name service. PID = " $pid_of_nsport
                 sudo kill -9 $pid_of_nsport
             else
+                debug "Killing name service. PID = " $pid_of_nsport
                 echo $PASSWORD | sudo -S kill -9 $pid_of_nsport
             fi
-            echo "$pname (pid: $pid_of_nsport) are killed"
-            debug "$pname (pid: $pid_of_nsport) are killed"
+            echo "$PNAME (pid: $pid_of_nsport) is killed"
+            debug "$PNAME (pid: $pid_of_nsport) is killed"
             delete_omninames_files
             return 0
         fi
-        echo "No running $pname found. The process using the port $PORT "
+        echo "No running $PNAME found. The process using the port $PORT "
         echo "cannot be estimated. Arboting"
         exit 1
     fi
@@ -334,34 +408,54 @@ stop_existing_ns()
 #
 start_omninames()
 {
-    debug "---start_omninames"
-    echo 'Starting omniORB omniNames: '$hostname':'$PORT
+    debug "start_omninames()"
+    debug "---start_omninames---"
+    echo  'Starting omniORB omniNames: '$HOSTNAME':'$PORT
+    debug 'Starting omniORB omniNames: '$HOSTNAME':'$PORT
     delete_omninames_files
-    if test "x$PASSWORD" = "x" ; then
-        $cosnames -start $PORT -logdir $currdir &
+
+    debug "$COSNAMES -start $PORT -logdir $LOGDIR"
+    $COSNAMES -start $PORT -logdir $LOGDIR &
+    if test $? != 0; then
+        debug "omniNames start failed."
+        return 1
     else
-        echo $PASSWORD | sudo -S $cosnames -start $PORT -logdir $currdir &
+        debug "omniNames started."
     fi
-    ret=$!
+
+    pid=$!
+    debug "PID: " $!
+    debug "Return: " $?
+
+    # After omniNames started, if another omniNames already working in
+    # the same port, omniNames fails and terminated after a few seconds.
+    # So one second sleep is necessary.
     sleep 1
-    debug "return code of omniNames:" $ret
-    omnip=`ps $! | wc -l`
-    if test $omnip -gt 1; then
-        echo "omniNames properly started"
+
+    # check if the omniNames strill working
+    omnipid=`ps $pid | grep $pid | grep -v grep | awk '{print $1;}'`
+    debug   "omnipid: "$omnipid
+    if test "x$omnipid" != "x" ; then
+        debug "omniNames properly started"
+        echo  "omniNames properly started"
         exit 0
+    else
+        debug "omniNames start failed."
+        echo  "omniNames start failed."
     fi
     return 1 
 }
 
 omniname()
 {
+    debug "omniname()"
     # Check cosname variable is properly set
     check_cosname
 
     # Starting omniNames 
     if test "x$STOPNAMESERVICE" = "x" ; then
-      start_omninames
-      echo "omniNames was not properly started."
+        start_omninames
+        echo "omniNames was not properly started."
     fi
 
     # Stopping existing name service
@@ -369,8 +463,8 @@ omniname()
 
     # Try to start omniNames again
     if test "x$STOPNAMESERVICE" = "x" ; then
-      start_omninames
-      echo "omniNames was not properly started."
+        start_omninames
+        echo "omniNames was not properly started."
     fi
 }
 
@@ -385,12 +479,13 @@ omniname()
 #
 start_taonames()
 {
-    if test ! -f $cosnames ; then
+    debug "start_taonames()"
+    if test ! -f $COSNAMES ; then
 	echo "TAO Naming_Service not found. Aborting."
 	exit 1
     fi
-    echo 'Starting TAO Naming_Service: '$hostname':'$PORT
-    $cosnames -m 0 -ORBListenEndpoints iiop://:$PORT &
+    echo 'Starting TAO Naming_Service: '$HOSTNAME':'$PORT
+    $COSNAMES -m 0 -ORBListenEndpoints iiop://:$PORT &
     ret=$!
     sleep 1
     debug "return code of TAO Naming_Service:" $ret
@@ -405,6 +500,7 @@ start_taonames()
 
 taonames()
 {
+    debug "taonames()"
     # Check cosname variable is properly set
     check_cosname
 
@@ -427,7 +523,7 @@ taonames()
 #------------------------------------------------------------
 get_opt $@
 
-case $orb in
+case $ORB in
 	omniORB)
 		omniname
 		;;
